File: Main/Ragtime.Common/DataService/DataService.cs
/*
1. Этот файл разработан так, чтобы быть включенным непосредственно в клиентский проект. Он содержит все, чтобы избавить клиентский проект от лишних зависимостей.
2. Проект Ragtime не использует этот файл, просто компилирует, чтобы проверить хотя бы на дым.
*/
namespace Ragtime.DataService {
  using System;
  using System.Threading.Tasks;
  using System.Collections.Generic;
  using System.Net;
  using System.IO;
  using Newtonsoft.Json;
  using Newtonsoft.Json.Linq;
  using Newtonsoft.Json.Serialization;


  /// <summary>Клиент для доступа к веб-приложению Ragtime</summary>
  public partial class Client {


    /// <summary>Описание ошибки</summary>
    public class Fault {

      /// <summary>Сообщение</summary>
      [JsonProperty("message")]
      public string Message;

      /// <summary>Дополнительные сведения</summary>
      [JsonProperty("details")]
      public string Details;

      /// <summary>Внутренняя ошибка или предназначена пользователю?</summary>
      [JsonProperty("isInternal")]
      public bool IsInternal;

      /// <summary> Код ошибки </summary>
      [JsonProperty("errorId")]
      public string ErrorId;
    }


    /// <summary>Делегат событий</summary>
    public delegate void Notification();


    /// <summary>Ошибка вызова</summary>
    public class FaultException: Exception {
      public FaultException(Fault fault) : base(fault.Message) {
        Fault = fault;
      }

      /// <summary>Подробности ошибки</summary>
      public readonly Fault Fault;
    }

    public event Notification OnStartRequest;
    public event Notification OnEndRequest;


    /// <summary>Адрес сервера (без пути). Например, http://some-server.com</summary>
    public string ServerAddress {
      get { return _ServerAddress; }
      set {
        if(value != null) {
          value = value.Trim();
          while(value.EndsWith("/"))
            value = value.Substring(0, value.Length-1);
          if(value == "")
            value = null;
        }
        _ServerAddress = value;
      }
    }
    private string _ServerAddress;

    public string ProxyAddress;
    public string ProxyUserName;
    public string ProxyPassword;

    /// <summary>Идентификатор приложения. Генерируется автоматически, но можно задать и явно</summary>
    public Guid AppInstanceId;

    /// <summary>Токен авторизации</summary>
    public string User;

    /// <summary>Вызываем сервер</summary>
    public Task<T> Call<T>(string service, string method, object context, object args) {
      bool runNow;
      lock(_Lock) {
        runNow = _Batch == null;
        if(_Batch == null)
          BeginBatch();
        var result = NewCall<T>(service, method, context, args);
        if(runNow)
          Forget(SendBatch());
        return result;
      }
    }

    /// <summary>Вызываем сервер</summary>
    public Task<T> Call<T>(string service, string method, object args) {
      return Call<T>(service, method, null, args);
    }

    /// <summary>Вызываем сервер</summary>
    public Task Call(string service, string method, object context, object args) {
      return Call<object>(service, method, context, args);
    }

    /// <summary>Вызываем сервер</summary>
    public Task Call(string service, string method, object args) {
      return Call<object>(service, method, null, args);
    }

    /// <summary>Начинаем пакет. После вызоваэтого метода ни один вызов DataService-а не пойдет на сервер до момента вызова sendBatch</summary>
    public void BeginBatch() {
      if(_Batch == null)
        _Batch = new List<DataService.Client.CallDto>();
    }

    /// <summary>Отправляем накопленный пакет на сервер</summary>
    public async Task SendBatch() {
      Dictionary<int, Callback> callbacks;
      CallDto[] batch;

      lock(_Lock) {
        if(AppInstanceId == Guid.Empty)
          AppInstanceId = Guid.NewGuid();

        callbacks = _Callbacks;
        _Callbacks = new Dictionary<int, Callback>();

        batch = _Batch.ToArray();
        _Batch = null;
      }

      try {
        var httpRequest = CreateDataRequest();
        using(var stream = httpRequest.GetRequestStream())
        using(var sWriter = new StreamWriter(stream))
        using(var jWriter = new JsonTextWriter(sWriter)) {
          var request = new RequestDto() {
            timezoneOffset = -(int)DateTimeOffset.Now.Offset.TotalMinutes,
            appInstanceId  = AppInstanceId.ToString(),
            user           = User,
            items          = batch,
          };
          Serializer.Serialize(jWriter, request);
          jWriter.Flush();
        }

        HttpWebResponse response;
        OnStartRequest?.Invoke();
        try {
          response = await httpRequest.GetResponseAsync() as HttpWebResponse;
        }
        finally {
          OnEndRequest?.Invoke();
        }

        if(response.StatusCode != HttpStatusCode.OK) {
          var e = new Exception($"Сервер по адресу {ServerAddress} вернул ошибку {response.StatusCode}: {response.StatusDescription}");
          foreach(var id in callbacks.Keys) {
            callbacks[id].Answer.SetException(e);
            callbacks.Remove(id);
          }
        }

        using(var stream = response.GetResponseStream())
        using(var sReader = new StreamReader(stream))
        using(var jReader = new JsonTextReader(sReader)) {
          ParseResponse(JToken.ReadFrom(jReader) as JObject, callbacks);
        }
      }

      catch(Exception e) {
        foreach(var callback in callbacks.Values)
          callback.Answer.SetException(e);
      }

      finally {
        foreach(var callback in callbacks.Values) {
          // Этот код не должен исполняться: к этому моменту список callbacks должен быть пуст
          callback.Answer.SetException(new Exception("Неожиданная ошибка"));
        }
      }
    }

    /// <summary>Получаем поток указанного BLOB-а</summary>
    public async Task<Stream> GetBlob(Guid id) {
      var request = CreateBlobRequest(id);
      var response = await request.GetResponseAsync();
      return response.GetResponseStream();
    }

    /// <summary>Создаем новый запрос и помещаем его в список "list"</summary>
    private Task<T> NewCall<T>(string service, string method, object context, object args) {
      var call = new CallDto() {
        id      = _Callbacks.Count + 1,
        service = service,
        method  = method,
        context = context,
        args    = args,
      };
      _Batch.Add(call);

      var callback = new Callback<T>();
      _Callbacks[call.id] = callback;
      return callback.Result;
    }

    private HttpWebRequest CreateDataRequest() {
      if(ServerAddress == null)
        throw new Exception("Адрес сервера не указан");

      var result = WebRequest.Create(ServerAddress + "/$data") as HttpWebRequest;
      result.UserAgent = "Ragtime.DataService.Client";
      result.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
      result.Proxy = GetProxy();
      result.Timeout = 60 * 5 * 1000;
      result.Method = "POST";
      result.ContentType = "application/json; charset=utf-8";

      return result;
    }

    private HttpWebRequest CreateBlobRequest(Guid id) {
      if(ServerAddress == null)
        throw new Exception("Адрес сервера не указан");

      var result = WebRequest.Create($"{ServerAddress}/ragtime/blob?id={id}") as HttpWebRequest;
      result.UserAgent = "Ragtime.DataService.Client";
      result.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
      result.Proxy = GetProxy();
      result.Timeout = 60 * 5 * 1000;
      result.Method = "GET";

      return result;
    }

    private IWebProxy GetProxy() {
      if(_HasProxy == null) {
        if(ProxyAddress != null) {
          _HasProxy = true;
          _Proxy = new WebProxy(ProxyAddress);
          if(ProxyUserName != null)
            _Proxy.Credentials = new NetworkCredential(ProxyUserName, ProxyPassword);
        }
        else {
          _HasProxy = false;
          _Proxy = HttpWebRequest.DefaultWebProxy;
        }
      }
      return _Proxy;
    }
    private bool?     _HasProxy;
    private IWebProxy _Proxy;

    private void Forget(Task task) {
      // Ну да, ничего не делаем. Forget же...
    }

    private void ParseResponse(JObject root, Dictionary<int, Callback> callbacks) {
      if(root.TryGetValue("fault", out var jRootFault)) {
        var fault = jRootFault.ToObject<Fault>(Serializer);
        foreach(var callback in callbacks.Values)
          callback.Answer.SetException(new FaultException(fault));
        return;
      }

      root.TryGetValue("authenticated", out var authenticated);
      if(!(bool)authenticated)
        User = null;

      if(root.TryGetValue("items", out var items)) {
        foreach(JObject item in items as JArray) {
          var callId = (int)item["callId"];

          Callback callback;
          if(callbacks.TryGetValue(callId, out callback)) {
            if(item.TryGetValue("fault", out var jCallFault)) {
              var fault = jCallFault.ToObject<Fault>(Serializer);
              callback.Answer.SetException(new FaultException(fault));
            }
            else {
              if(item.TryGetValue("result", out var result)) {
                callback.Answer.SetResult(result);
              }
              else
                callback.Answer.SetResult(null);
            }
            callbacks.Remove(callId);
          }
        }
      }
    }


    /// <summary>Список вызовов, ожидающих передачи</summary>
    private List<CallDto> _Batch;

    /// <summary>Ответы на вызовы Call(). Ключ: CallId</summary>
    private Dictionary<int, Callback> _Callbacks = new Dictionary<int, Callback>();

    /// <summary>Точка блокировки</summary>
    private object _Lock = new object();


    /// <summary>Информация о вызове (нетипизированная)</summary>
    private class Callback {

      /// <summary>Сюда мы поместим ответ от сервера</summary>
      public readonly TaskCompletionSource<JToken> Answer = new TaskCompletionSource<JToken>();
    }


    /// <summary>Информация о вызове (типизированная)</summary>
    private class Callback<T>: Callback {

      /// <summary>Эта задача разрешится, как только получит ответ от сервера и десериализует его</summary>
      public Task<T> Result { get; private set; }

      public Callback() {
        Result = Task.Run<T>(async () => {
          var data = await Answer.Task;
          if(data == null)
            return default(T);
          else
            return data.ToObject<T>(Client.Serializer);
        });
      }
    }


    /// <summary>Dto запроса</summary>
    private class RequestDto {

      /// <summary>Часовой пояс</summary>
      public long timezoneOffset;

      /// <summary>Id приложения</summary>
      public string appInstanceId;

      /// <summary>Токен авторизации</summary>
      public string user;

      /// <summary>Вызовы</summary>
      public CallDto[] items;
    }

    /// <summary>Вызов (элемент пачки)</summary>
    private class CallDto {
      public int id;
      public string service;
      public string method;
      public object context;
      public object args;
    }
  }


  partial class Client {

    private static JsonSerializer Serializer => new JsonSerializer() { ContractResolver = ContractResolver.Instance };

    private class ContractResolver: DefaultContractResolver {
      public static ContractResolver Instance = new ContractResolver();

      protected override JsonConverter ResolveContractConverter(Type objectType) {
        foreach(var c in _Converters)
          if(c.CanConvert(objectType))
            return c;
        return null;
      }
      private static JsonConverter[] _Converters = new JsonConverter[] {
        new GuidConverter(),
      };
    }

    private class GuidConverter: JsonConverter {
      public override bool CanConvert(Type objectType) {
        return objectType == typeof(Guid);
      }
      public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer) {
        if(!Guid.TryParse((string)reader.Value, out var result))
          result = Guid.Empty;
        return result;
      }
      public override void WriteJson(JsonWriter writer, object _value, JsonSerializer serializer) => writer.WriteValue(_value);
    }
  }
}

File: Main/Ragtime.Common/DataService/DataService.ts
//summary: Низкоуровневый протокол взаимодействия между клиентской стороной (браузером) и сервером 
//alias:   Ragtime.DataService
//bundle:  true

import { IMap } from "Ragtime.Types";
import { TypedCallbacks as Callbacks } from "Ragtime.Callbacks";


/** Описание ошибки */
export interface Fault {

  /** Сообщение */
  message: string;

  /** Дополнительные сведения */
  details?: string;

  /** Внутренняя ошибка или предназначена пользователю ? */
  isInternal?: boolean;

  /** Код ошибки */
  errorId?: string;
}


/** Результат вызова */
export interface CallResult {

  /** Контекст (то есть this), если имеется */
  context?: any;

  /** Визуальное состояние модели (набор флагов enabled, visible, etc) */
  modelState?: any;

  /** Непосредственно результат вызова */
  result?: any;
}



/** Сигнатура обработчика *Request */
export type RequestHandler = (request: Request) => any;

/** Сигнатура обработчика *Response */
export type ResponseHandler = (response: Response) => any;

/** Сигнатура обработчика *Fault */
export type FaultHandler = (fault: Fault) => any;


/** Регистрируем обработчик, который будет вызываться перед вызовом сервера */
export function addBeforeRequest(handler: RequestHandler) {
  beforeRequest.add(handler);
}
var beforeRequest = new Callbacks<RequestHandler>();

/** Регистрируем обработчик, который будет вызываться сразу после получения ответа от сервера */
export function addAfterResponse(handler: ResponseHandler) {
  afterResponse.add(handler);
}
var afterResponse = new Callbacks<ResponseHandler>();

/** Регистрируем обработчик, который будет вызываться после вызова сервера - неважно, успешного или нет */
export function addAfterRequest(handler: RequestHandler) {
  afterRequest.add(handler);
}
var afterRequest = new Callbacks<RequestHandler>();

/** Регистрируем обработчик, который будет вызываться при ошибке */
export function addWhenFault(handler: FaultHandler) {
  whenFault.add(handler);
}
var whenFault = new Callbacks<FaultHandler>();


/** Вызываем сервер без контекста (вызов "статического" метода) */
export function call<Result = any>(service: string, method: string, args: any, background: boolean = null): Promise<Result> {
  let runNow = !batched;
  if(!batched)
    beginBatch();
  let result = newCall(service, method, null, args, _ => batched.push(_), background);
  if(runNow)
    sendBatch();
  return new Promise<Result>((resolve, reject) => {
    result.then(
      _ => resolve(_.result),
      _ => reject(_));
  });
}

/** Вызываем сервер с контекстом (вызов "метода экземпляра") */
export function callWithContext(service: string, method: string, context: any, args: any, background: boolean = null): Promise<CallResult> {
  let runNow = !batched;
  if(!batched)
    beginBatch();
  let result = newCall(service, method, context, args, _ => batched.push(_), background);
  if(runNow)
    sendBatch();
  return result;
}

/** Вызываем сервер отложенно. Вызов будет сделан либо вместе с первым неотложенным вызовом, либо при вызове batch-а */
export function postponeCall<Result>(service: string, method: string, args: any, priority: boolean = false, background: boolean = null): Promise<Result> {
  return new Promise<Result>((resolve, reject) => {
    let add 
      = priority 
        ? (_: Call) => postponed.splice(0,0,_) 
        : (_: Call) => postponed.push(_)
      ;
    newCall(service, method, null, args, add, background).then(
      _ => resolve(_.result),
      _ => reject(_));
  });
}
  
/** Начинаем пакет. После вызоваэтого метода ни один вызов DataService-а не пойдет на сервер до момента вызова sendBatch */
export function beginBatch(background: boolean = null) {
  if(!batched)
    batched = [];
  if(background !== null)
    defaultBackground = background;
}

/** Выполняем тело в контексте пакета */
export async function batch(body: () => any): Promise<void> {
  beginBatch();
  try {
    body();
  }
  finally {
    await sendBatch();
  }
}

/** Отправляем накопленный пакет на сервер */
export async function sendBatch(): Promise<void> {
  let done: (value?: any) => any;
  let result = new Promise<void>((resolve, reject) => done = resolve);

  let request: Request = {
    timezoneOffset: new Date().getTimezoneOffset(),
    items: null,
    background: false,
  };

  request.items = (batched || []).concat(postponed);
  request.background = request.items.every(_ => _.background);
  beforeRequest.fire(request); // Это довольно важно - вызвать эти обработчики здесь. Допускаю, что какие-то из этих обработчиков вызовут postponeCall() или call().

  request.items = (batched || []).concat(postponed);
  request.background = request.items.every(_ => _.background);

  let cbcks = callbacks;
  batched = null;
  postponed = [];
  callbacks = {};
  defaultBackground = false;

  try {
    let fetchResult = await fetch("/$data", {
      method: "POST",
      cache: "no-cache",
      body: JSON.stringify(request),
      headers: {
        "Content-Type": "application/json; charset=utf-8",
      },
      credentials: "include",
    });

    if(fetchResult.ok) {
      let response: Response = await fetchResult.json();
      afterResponse.fire(response);
      if(response.fault) {
        handleError(response.fault, request.background);
        return;
      }
      try {
        for(let item of response.items) {
          let call = cbcks[item.callId];
          if(call) {
            if(!item.fault) {
              delete item.callId;
              call.resolve(item);
            }
            else {
              call.reject(item.fault.message);
              if(!call.background)
                whenFault.fire(item.fault);
            }
            try {
              await call.promise;
            }
            catch {
            }
          }
        }
      }
      finally {
        done();
        afterRequest.fire(request);
      }
    }

    else 
      handleError(createFault(fetchResult.status, fetchResult.statusText, await fetchResult.text()), request.background);
  }
  catch(e) {
    handleError(createFault(0, "Неожиданная ошибка", e.message), request.background);
  }

  return result;


  async function handleError(fault: Fault, background: boolean) {
    try {
      for(let id in cbcks) {
        let callback = cbcks[id];
        callback.reject(null);
        try {
          await callback.promise;
        }
        catch {
        }
      }
      if(!background)
        whenFault.fire(fault);
    }
    finally {
      done();
      afterRequest.fire(request);
    }
  }
}


/** Dto пачки */
export interface Request {

  /** Часовой пояс */
  timezoneOffset: number;

  /** Id приложения */
  appInstanceId?: string;

  /** Пользователь (зашифрованный токен) */
  user?: string;

  /** Данные сессии профилирования */
  profilingSession?: any;

  /** Вызовы */
  items: Call[];

  /** Фоновый вызов: о ходе вызова пользователя не оповещаем */
  background: boolean;
}

/** Вызовы */
export interface Call {
  id: number;
  service: string;
  method: string;
  context: any;
  args: any;
  background: boolean,
}

export interface Response {
  authenticated: boolean;
  dbUnderMaintenance: boolean;
  items: Answer[];
  fault: Fault;
}

interface Answer {
  callId: number;
  context: any;
  result: any;
  fault: Fault;
}

interface Callback {
  promise?: Promise<CallResult>;
  resolve?: (value?: any) => any;
  reject?:  (value?: any) => any;
  background?: boolean;
}

/** Создаем новый запрос и помещаем его в список "list" */
function newCall(service: string, method: string, context: any, args: any, add: (call: Call) => any, background: boolean): Promise<CallResult> {
  if(typeof(background) !== "boolean")
    background = defaultBackground;
  let call: Call = {
    id: Object.keys(callbacks).length + 1,
    service,
    method,
    context,
    args,
    background
  };
  add(call);

  let callback: Callback = {};
  callbacks[call.id] = callback;
  callback.promise = new Promise<any>((resolve, reject) => {
    callback.resolve = resolve;
    callback.reject  = reject;
  });
  callback.background = background;
  return callback.promise;
}

export function createFault(status: number, statusText: string, responseText: string): Fault {
  let result: Fault = null;
  if(statusText && statusText != "error") {
    result = {
      message: "Произошла неожиданная ошибка: " + statusText,
    }
  }
  if(!result) {
    if(responseText) {
      try {
        result = JSON.parse(responseText) as Fault;
      }
      catch(e) {
      }
    }
  }
  if(!result) {
    result = {
      message: responseText,
    };
    if(!result.message)
      result.message = "Произошла неожиданная ошибка: " + status;
  }
  return result;
}


/** Список вызовов, которые ожидают передачи в пачке */
var batched: Call[] = null;

var defaultBackground: boolean = false;

/** Список отложенных вызовов */
var postponed: Call[] = [];

/** Список ответов */
var callbacks: IMap<Callback> = {};
